Snakefile

```diff
@@ -0,0 +1,377 @@
+
+rule upload:
+    input:
+        **files_to_upload
+    output:
+        touch(f"data/{database}/upload.done")
+    params:
+        quiet = "" if send_notifications else "--quiet",
+        s3_bucket = config["s3_dst"]
+    run:
+        for remote, local in input.items():
+            shell("./bin/upload-to-s3 {params.quiet} {local:q} {params.s3_bucket:q}/{remote:q}")
+
+
+rule trigger_preprocessing_pipeline:
+    message: "Triggering nextstrain/ncov rebuild action (via repository dispatch)"
+    input:
+        f"data/{database}/upload.done"
+    output:
+        touch(f"data/{database}/trigger-preprocessing.done")
+    params:
+        dispatch_type = f"preprocess-{database}",
+        token = os.environ.get("PAT_GITHUB_DISPATCH", "")
+    run:
+        headers = {
+                'Content-type': 'application/json',
+                'authorization': f"Bearer {params.token}",
+                'Accept': 'application/vnd.github.v3+json'}
+        data = {"event_type": dispatch_type}
+        print(f"Triggering ncov preprocessing GitHub action via repository dispatch type: {params.dispatch_type}")
+        response = requests.post("https://api.github.com/repos/nextstrain/ncov/dispatches", headers=headers, data=json.dumps(data))
+        response.raise_for_status()
+
+################################################################
+################################################################
+
+# A helpful list of environment variables in use by various scripts
+env_variables = {
+    "AWS_DEFAULT_REGION": "Required for S3 access",
+    "AWS_ACCESS_KEY_ID": "Required for S3 access",
+    "AWS_SECRET_ACCESS_KEY": "Required for S3 access",
+    "GITHUB_RUN_ID": "Included in slack notification message (optional)",
+    "SLACK_TOKEN": "Required for sending slack notifications",
+    "SLACK_CHANNELS": "Required for sending slack notifications",
+    "PAT_GITHUB_DISPATCH": "Required for triggering preprocessing GitHub actions",
+    "GISAID_API_ENDPOINT": "Required for GISAID API access",
+    "GISAID_USERNAME_AND_PASSWORD": "Required for GISAID API access"
+}
+
+onstart:
+    print(f"Pipeline starting.")
+    print(f"Source s3 bucket: {config['s3_src']}, destination: {config['s3_dst']}")
+    print("Environment variables present:")
+    for var, description in env_variables.items():
+        print(f"\t${{{var}}}: " + ("YES" if os.environ.get(var, "") else "NO") + f"({description})")
+    if send_notifications:
+        message="🥗 GISAID ingest" if database=="gisaid" else "🥣 GenBank ingest"
+        shell(f"./bin/notify-on-job-start \"{message}\"")
+
+onsuccess:
+    message = "✅ This pipeline has successfully finished 🎉"
+    print(message)
+    shell("./bin/clean")
+
+onerror:
+    message = "❌ This pipeline has FAILED 😞. Please see linked thread for more information."
+    print(message)
+    print("P.S. Because this failed, we did not run './bin/clean' but you may wish to run this now.")
```

Comment:
Does not running clean here on failure mean that remote jobs may try to upload several large files when they fail?

If so, would it be better to only skip clean on local runs?

Reply:
Are you here referring to the `nextstrain build` behavior on AWS where it'll upload the entire working directory as a (zipped) asset to S3?

> If so, would it be better to only skip clean on local runs

Sure thing. My inclination would be to have an explicit opt-in config param here, such as `--config keep_all_files=True`, do you have an alternative method you'd prefer?

Reply:
Yes, that `nextstrain build` behaviour.

Opt-in config flag is fine by me! (Though it may mean someone debugging a local run will forget too set it and then have to re-run with it set to inspect failed outputs.)

Reply:
Actually, I'm going to merge as-is for expediency as the current behavior preserves the previous (bash script) behavior where `./bin/clean` was skipped if there was an error in any previous step. (The one exception being if the GitHub action trigger were to fail.) I do agree that clean-up of files is important and we should audit this as part of #240.

Reply:
Implemented in 45dcea6 and i'll also note this on #240.